polyfills/requestIdleCallback/polyfill.js

```diff
@@ -0,0 +1,306 @@
+                // It appears to be (now + (frame rate) * 2).
+                var futureRafTime = rafTime - activeFrameTime > performance.now();
+                if (futureRafTime) {
+                    rafTime = rafTime - (activeFrameTime * 2);
+                }
+                // Calculate the frame rate.
+                var nextFrameTime = rafTime - frameDeadline + activeFrameTime;
+                if (nextFrameTime < activeFrameTime && previousFrameTime < activeFrameTime) {
+                    if (nextFrameTime < 8) {
+                        // Defensive coding. We don't support higher frame rates than 120hz.
+                        // If we get lower than that, it is probably a bug.
+                        nextFrameTime = 8;
+                    }
+                    // If one frame goes long, then the next one can be short to catch up.
+                    // If two frames are short in a row, then that's an indication that we
+                    // actually have a higher frame rate than what we're currently optimizing.
+                    // We adjust our heuristic dynamically accordingly. For example, if we're
+                    // running on 120hz display or 90hz VR display.
+                    // Take the max of the two in case one of them was an anomaly due to
+                    // missed frame deadlines.
+                    activeFrameTime = Math.max(previousFrameTime, nextFrameTime);
+                } else {
+                    previousFrameTime = nextFrameTime;
+                }
+                // Update the deadline we have to run idle callbacks.
+                frameDeadline = rafTime + activeFrameTime;
+                // Schedule idle callback work.
+                scheduleIdleWork();
+            });
+
+            // If the animation frame is not called, for example if the tab is in the
+            // background, schedule work should still be run. So set a timeout as a
+            // fallback.
+            scheduledAnimationFrameTimeout = setTimeout(function () {
+                // Cancel the animation frame which failed to run timely.
+                cancelAnimationFrame(scheduledAnimationFrameId);
+                scheduledAnimationFrameId = undefined;
+                // Update the deadline we have to run idle callbacks.
+                frameDeadline = performance.now() + 50;
+                // Run callbacks.
+                scheduleIdleWork();
+            }, 100);
+        }
+    }
+
+    function processIdleCallbacks () {
+        isIdleScheduled = false;
+        isCallbackRunning = true;
+
+        // Move timed-out callbacks from the scheduled array.
+        timedOutCallbacks = scheduledCallbacks.filter(function (callbackObject) {
+            if (callbackObject.options.timeout !== undefined) {
+                var deadline = getDeadline(callbackObject);
+                return deadline.didTimeout;
+            }
+            return false;
+        }, []);
```

Comment:
I don't think we need to set the `thisArg` of Array.prototype.filter to an array.
```suggestion
        });
```